service/history/shard/controller_impl.go

```diff
@@ -208,6 +211,11 @@ func (c *ControllerImpl) CloseShardByID(shardID int32) {
 }
 
 func (c *ControllerImpl) shardClosedCallback(shard *ContextImpl) {
+	defer func() {
+		if c.shardClosed != nil {
+			c.shardClosed(shard)
+		}
+	}()
```

Comment:
what is this dong? It seems too much just for test.

Reply:
It's a callback for when the shard is closed. We need to know when it's closed in order to test without relying on a sleep. I'll remove it though since this is going to a context test.